File: src/algorithms/strings/pattern-matching/z-algorithm/sources/z-algorithm_test.rs
include!("z-algorithm.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pattern_at_start() {
        assert_eq!(z_algorithm("ABCDEF", "ABC"), 0);
    }

    #[test]
    fn test_pattern_in_middle() {
        assert_eq!(z_algorithm("AABXAABXCAABXAABXAY", "AABXAAB"), 0);
    }

    #[test]
    fn test_pattern_near_end() {
        assert_eq!(z_algorithm("XYZAABXAAB", "AABXAAB"), 3);
    }

    #[test]
    fn test_pattern_at_end() {
        assert_eq!(z_algorithm("XYZABC", "ABC"), 3);
    }

    #[test]
    fn test_pattern_not_found() {
        assert_eq!(z_algorithm("ABCDEFG", "XYZ"), -1);
    }

    #[test]
    fn test_single_char_found() {
        assert_eq!(z_algorithm("HELLO", "L"), 2);
    }

    #[test]
    fn test_single_char_not_found() {
        assert_eq!(z_algorithm("HELLO", "Z"), -1);
    }

    #[test]
    fn test_empty_pattern() {
        assert_eq!(z_algorithm("HELLO", ""), 0);
    }

    #[test]
    fn test_text_equals_pattern() {
        assert_eq!(z_algorithm("ABCD", "ABCD"), 0);
    }

    #[test]
    fn test_pattern_longer_than_text() {
        assert_eq!(z_algorithm("AB", "ABCD"), -1);
    }

    #[test]
    fn test_repeated_characters() {
        assert_eq!(z_algorithm("AAAAAB", "AAAB"), 2);
    }

    #[test]
    fn test_first_of_multiple() {
        assert_eq!(z_algorithm("ABABABAB", "ABAB"), 0);
    }
}

File: src/algorithms/strings/pattern-matching/z-algorithm/__tests__/z-algorithm_test.go
package main

import "testing"

func TestZAlgorithmPatternAtStart(t *testing.T) {
	if zAlgorithm("ABCDEF", "ABC") != 0 {
		t.Error("expected 0")
	}
}

func TestZAlgorithmPatternInMiddle(t *testing.T) {
	if zAlgorithm("AABXAABXCAABXAABXAY", "AABXAAB") != 0 {
		t.Error("expected 0")
	}
}

func TestZAlgorithmPatternNearEnd(t *testing.T) {
	if zAlgorithm("XYZAABXAAB", "AABXAAB") != 3 {
		t.Error("expected 3")
	}
}

func TestZAlgorithmPatternAtEnd(t *testing.T) {
	if zAlgorithm("XYZABC", "ABC") != 3 {
		t.Error("expected 3")
	}
}

func TestZAlgorithmPatternNotFound(t *testing.T) {
	if zAlgorithm("ABCDEFG", "XYZ") != -1 {
		t.Error("expected -1")
	}
}

func TestZAlgorithmSingleCharFound(t *testing.T) {
	if zAlgorithm("HELLO", "L") != 2 {
		t.Error("expected 2")
	}
}

func TestZAlgorithmSingleCharNotFound(t *testing.T) {
	if zAlgorithm("HELLO", "Z") != -1 {
		t.Error("expected -1")
	}
}

func TestZAlgorithmEmptyPattern(t *testing.T) {
	if zAlgorithm("HELLO", "") != 0 {
		t.Error("expected 0 for empty pattern")
	}
}

func TestZAlgorithmTextEqualsPattern(t *testing.T) {
	if zAlgorithm("ABCD", "ABCD") != 0 {
		t.Error("expected 0")
	}
}

func TestZAlgorithmPatternLongerThanText(t *testing.T) {
	if zAlgorithm("AB", "ABCD") != -1 {
		t.Error("expected -1")
	}
}

func TestZAlgorithmRepeatedChars(t *testing.T) {
	if zAlgorithm("AAAAAB", "AAAB") != 2 {
		t.Error("expected 2")
	}
}

func TestZAlgorithmFirstOfMultiple(t *testing.T) {
	if zAlgorithm("ABABABAB", "ABAB") != 0 {
		t.Error("expected 0")
	}
}
